numpy/core/src/multiarray/compiled_base.c

```diff
@@ -825,29 +834,44 @@ static int sequence_to_arrays(PyObject *seq,
     }
 
     for (i = 0; i < count; ++i) {
+        PyArray_Descr *dtype = NULL;
         PyObject *item = PySequence_GetItem(seq, i);
         if (item == NULL) {
-            while (--i >= 0) {
-                Py_DECREF(op[i]);
-                op[i] = NULL;
-            }
-            return -1;
+            goto fail;
         }
-
-        op[i] = (PyArrayObject *)PyArray_FROM_O(item);
-        if (op[i] == NULL) {
-            while (--i >= 0) {
-                Py_DECREF(op[i]);
-                op[i] = NULL;
+        if (PyArray_DTypeFromObject(item, NPY_MAXDIMS, &dtype) < 0) {
+            Py_DECREF(item);
+            goto fail;
+        }
+        if (dtype == NULL) {
+            if (PySequence_Check(item) && PySequence_Size(item) == 0) {
+                PyErr_SetString(PyExc_TypeError, EMPTY_SEQUENCE_ERR_MSG);
             }
             Py_DECREF(item);
-            return -1;
+            goto fail;
         }
-
+        if (!(PyDataType_ISINTEGER(dtype) || PyDataType_ISBOOL(dtype))) {
+            PyErr_SetString(PyExc_TypeError, NON_INTEGRAL_ERROR_MSG);
+            Py_DECREF(dtype);
+            Py_DECREF(item);
+            goto fail;
+        }
+        op[i] = (PyArrayObject *)PyArray_FromAny(item, dtype, 0, 0, 0, NULL);
         Py_DECREF(item);
+        if (op[i] == NULL) {
+            Py_DECREF(dtype);
```

Comment:
I think this double-decrefs, `PyArray_FromAny` (probably) steals a reference

Reply:
fixed. It seems this comes up often enough I should have a pre-commit hook to find these.